### PR TITLE
Moved Responsibility for Creating Primitives From ROS Messages Out of Primitive Class

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Moved Responsibility for Creating Primitives From ROS Messages Out of Primitive Class

- created `createPrimitiveFromROSMsg` function in `primitive_factory.cpp`
- remove equivalent functionality from `Primitive`
- moved all relevant tests into `test/.../primitive_factory.cpp`
- added tests for `validatePrimitiveMessage` that probably should've been added a while back when we added exceptions

### Testing Done
CI (should) pass

### Resolved Issues
resolves 133
resolves 217

### Length Justification
Moved a lot of lines, and remove the spiderweb of dependencies on various primitives.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/test/primitive/primitive_factory.cpp [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        robot_id, destination, final_angle, final_speed, rpm, small_kick_allowed);
+
+    thunderbots_msgs::Primitive prim_message = dribble_prim.createMsg();
+
+    std::unique_ptr<Primitive> new_prim =
+        AI::Primitive::createPrimitiveFromROSMessage(prim_message);
+
+    // Since we have a `Primitive` and NOT a `MovePrimitive`, we have to check that the
+    // values are correct by getting them from the generic parameter array
+    std::vector<double> params  = new_prim->getParameters();
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/primitive/primitive_factory.cpp:293]: Just a thought, you can `dynamic_cast` to a `MovePrimitive` and then compare values more explicitly, or make use of equality operators (although we don't define these yet)

File: src/thunderbots/software/ai/primitive/primitive_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+#include "ai/primitive/directwheels_primitive.h"
+#include "ai/primitive/dribble_primitive.h"
+#include "ai/primitive/kick_primitive.h"
+#include "ai/primitive/move_primitive.h"
+#include "ai/primitive/movespin_primitive.h"
+#include "ai/primitive/pivot_primitive.h"
+#include "ai/primitive/stop_primitive.h"
+
+
+std::unique_ptr<::Primitive> AI::Primitive::createPrimitiveFromROSMessage(
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/primitive/primitive_factory.cpp:21]: Is the double colon necessary in front of `Primitive` here? (and everywhere else)?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Quick question but LGTM otherwise

File: src/thunderbots/software/test/primitive/primitive_factory.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/test/primitive/primitive_factory.cpp:293]: More of a thought for the future. Equality operators would be useful.

Comment (MathewMacDougall) [reply src/thunderbots/software/test/primitive/primitive_factory.cpp:293]: Made #377  and #378 to implement the operators in the future. This suggestion isn't really applicable right now (until those are done), so don't worry about it.

File: src/thunderbots/software/ai/primitive/primitive_factory.cpp (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/primitive_factory.cpp:21]: Yup, until we put `Primitive` in the `AI::Primitive` namespace, we need to explicitly state that we want `Primitive` from the global namespace.

Actually, you've pointed out the one case where it looks like we don't technically need to do this, but I'm inclined to leave it as-is right now to be consistent with the other instances.

File: src/thunderbots/software/test/primitive/primitive_factory.cpp (hunk 1 above)
Comment (garethellis0) [reply src/thunderbots/software/test/primitive/primitive_factory.cpp:293]: :+1: can we mark this comment as resolved?

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Pull request merged**